engine/kv_engine.cpp

```diff
@@ -1720,3 +1736,193 @@ Status KVEngine::RestoreDlistRecords(DLRecord *pmp_record) {
+          queue_ptr = queue_uptr.get();
+          {
+            std::lock_guard<std::mutex> lg{list_mu_};
+            queue_uptr_vec_.emplace_back(queue_uptr.release());
+          }
+
+          HashEntry hash_entry_collection;
+          HashEntry *p_hash_entry_collection = nullptr;
+          Status s = hash_table_->SearchForWrite(
+              hint_collection, collection_name, RecordType::QueueRecord,
+              &p_hash_entry_collection, &hash_entry_collection, nullptr);
+          kvdk_assert(s == Status::NotFound, "Logically impossible!");
+          hash_table_->Insert(
+              hint_collection, p_hash_entry_collection, RecordType::QueueRecord,
+              reinterpret_cast<uint64_t>(queue_ptr), HashOffsetType::Queue);
+        }
+      }
+    }
+  }
+
+  // Push
+  {
+    TimeStampType ts = get_timestamp();
+    switch (where) {
+    case QueueOpWhere::Left: {
+      queue_ptr->PushFront(ts, value);
+      return Status::Ok;
+    }
+    case QueueOpWhere::Right: {
+      queue_ptr->PushBack(ts, value);
+      return Status::Ok;
+    }
+    default: {
+      GlobalLogger.Error("Impossible!");
+      return Status::Abort;
+    }
+    }
+  }
+} catch (std::bad_alloc const &ex) {
+  return Status::PmemOverflow;
+}
+
+Status KVEngine::RestoreQueueRecords(DLRecord *pmp_record) {
+  switch (pmp_record->entry.meta.type) {
+  case RecordType::QueueRecord: {
+    Queue *queue_ptr = nullptr;
+    std::lock_guard<std::mutex> lg{list_mu_};
+    {
+      queue_uptr_vec_.emplace_back(
+          std::make_unique<Queue>(pmem_allocator_.get(), pmp_record));
+      queue_ptr = queue_uptr_vec_.back().get();
+    }
+
+    std::string collection_name = queue_ptr->Name();
+    HashTable::KeyHashHint hint_collection =
+        hash_table_->GetHint(collection_name);
+    std::unique_lock<SpinMutex>{*hint_collection.spin};
+
+    HashEntry hash_entry_collection;
+    HashEntry *p_hash_entry_collection = nullptr;
+    Status s = hash_table_->SearchForWrite(
+        hint_collection, collection_name, RecordType::QueueRecord,
+        &p_hash_entry_collection, &hash_entry_collection, nullptr,
+        true /* in recovery */);
+    hash_table_->Insert(
+        hint_collection, p_hash_entry_collection, RecordType::QueueRecord,
+        reinterpret_cast<uint64_t>(queue_ptr), HashOffsetType::Queue);
+    return Status::Ok;
+  }
+  case RecordType::QueueHeadRecord: {
+    kvdk_assert(pmp_record->prev == kNullPmemOffset &&
+                    checkDLRecordLinkageRight(pmp_record),
+                "Bad linkage found when RestoreDlistRecords. Broken head.");
+    return Status::Ok;
+  }
+  case RecordType::QueueTailRecord: {
+    kvdk_assert(pmp_record->next == kNullPmemOffset &&
+                    checkDLRecordLinkageLeft(pmp_record),
+                "Bad linkage found when RestoreDlistRecords. Broken tail.");
+    return Status::Ok;
+  }
+  case RecordType::QueueDataRecord: {
+    std::uint64_t offset_record =
```

Comment:
where is used `offset_record`? only for check ?

Reply:
removed. It was for debugging.